src/connector/src/macros.rs

```diff
@@ -56,12 +59,34 @@ macro_rules! impl_split {
                     .to_string()
             }
 
+            pub fn update(&self, start_offset: String) -> Self {
+                match self {
+                    $( Self::$variant_name(inner) => Self::$variant_name(inner.copy_with_offset(start_offset)), )*
+                }
+            }
+
             pub fn restore_from_bytes(split_type: String, bytes: &[u8]) -> Result<Self> {
                 match split_type.to_lowercase().as_str() {
                     $( $connector_name => <$split>::restore_from_bytes(bytes).map(Self::$variant_name), )*
                     other => Err(anyhow!("split type {} not supported", other)),
                 }
             }
+
+            $(
+                paste! {
+                pub fn [<get_ $variant_name:lower>](&self) -> Result<$split> {
```

Comment:
Bugen has previously imported an interesting crate https://crates.io/crates/enum-as-inner that might be helpful for this.

```
#[derive(Clone, Eq, PartialEq, Hash, EnumAsInner)]
pub enum ExprImpl {
    // ColumnRef(Box<BoundColumnRef>), might be used in binder.
    InputRef(Box<InputRef>),
    CorrelatedInputRef(Box<CorrelatedInputRef>),
    Literal(Box<Literal>),
    FunctionCall(Box<FunctionCall>),
    AggCall(Box<AggCall>),
    Subquery(Box<Subquery>),
}
```

Then you can use `expr.as_sub_query()` to access the internal variant.

Reply:
awesome! working on this